### ISSUE TITLE
Unable to override records with a paramerized Type?

### ISSUE THREAD

**Issue by ari-becker** [dhall-lang/dhall-lang#943]
### Good
attempting:
```dhall
{ foo = "foo" } ⫽ { foo = "bar" }
```
results in:
```dhall
{ foo = "bar"}
```
as expected.

### Problematic
attempting:
```dhall
let override =
      λ(type : Type) → λ(default : type) → λ(foo : type) → default ⫽ foo

in  override { foo : Text } { foo = "foo" } { foo = "bar" }
```
results in:
```text
Use "dhall --explain" for detailed errors

type : Type
default : type
foo : type

Error: You can only combine records

2│                                                            default ⫽ foo
3│ 

Example.dhall:2:60
```
, expecting the same output as in the Good section above.

**Comment by philandstuff** [comment]
What do you expect to happen if someone passes in `Natural` for the value of `type`? I’d say it has to be a type error, so this is working as expected. Can you explain what you’re trying to do?

This feels like it might be adjacent to row types, which were discussed in #434

**Comment by ari-becker** [comment]
> What do you expect to happen if someone passes in `Natural` for the value of `type`? I’d say it has to be a type error, so this is working as expected. Can you explain what you’re trying to do?
> 
> This feels like it might be adjacent to row types, which were discussed in #434

Would `5 // 9 === 9` really be so problematic?

**Comment by ari-becker** [comment]
> Can you explain what you're trying to do?

Small snippet:
```dhall
               config
              ⫽ { seal =
                    let seal =
                          Vault.config.Seal.AWS-KMS
                            Vault.config.Seal.Options.AWS-KMS::{
                            , region = Some parameters.aws.region
                            , access_key = Some
                                parameters.aws.credentials.access-key
                            , secret_key = Some
                                parameters.aws.credentials.secret-key
                            , kms_key_id = parameters.aws.kms.key-id
                            , endpoint = None Text
                            }

                    in  Some
                          ( merge
                              { Some =
                                    λ(default : Vault.config.Seal.Type)
                                  → default ⫽ seal
                              , None = seal
                              }
                              config.seal
                          )
                , telemetry =
                    let telemetry =
                          Vault.config.Telemetry::{
                          , common = Some Vault.config.Telemetry.Common::{
                            , disable_hostname = Some True
                            }
                          , prometheus = Some Vault.config.Telemetry.Providers.Prometheus::{
                            , retention_time = Some (Duration.Hours 24)
                            }
                          }

                    in  Some
                          ( merge
                              { Some =
                                    λ(default : Vault.config.Telemetry.Type)
                                  → default ⫽ telemetry
                              , None = telemetry
                              }
                              config.telemetry
                          )
                }
        }
    }
```
where `config.seal`, `config.telemetry` are `Optional` fields.

As you can see, the
```dhall
Some
                          ( merge
                              { Some =
                                    λ(default : Vault.config.Seal.Type)
                                  → default ⫽ seal
                              , None = seal
                              }
                              config.seal
                          )
```
is duplicated between `seal` and `telemetry`, and could (hypothetically) be refactored if `Vault.config.Seal.Type` and `Vault.config.Telemetry.Type` were parameterized.

From what I understand, the `with` syntax that's coming in `v15.0.0` will help, but I'm sure there are other cases.

**Comment by Gabriel439** [comment]
Like @philandstuff mentioned, this essentially requires support for row types.  The idea is that you could instead do something like:

```
let override =
        λ(row : Row)
      → λ(default : Record row)
      → λ(foo : Record row)
      → default ⫽ foo

in  override { foo : Text } { foo = "foo" } { foo = "bar" }
```

More generally, abstracting over `//` and `∧` operations requires support for row types.

**Comment by philandstuff** [comment]
This is quite old, I’m going to close it. Please reopen if this is still an issue for you.

**Comment by ari-becker** [comment]
Not re-opening, just for reference's sake, that this will be solved by #828 which is the currently open issue which deals with row types / row polymorphism.
